File: Problem solving with c++/Problem solving Level 2/Problem04.cpp
/*
Problem04 |===============================================

Write a program to print all perfect number From 1 to N 
order. 

Input : 
500

Output : 
6
28
496
==========================================================
*/



#include <iostream>
using namespace std;

// Enumeration to represent whether a number is Perfect or Not Perfect
enum enmPrefectNotPerfect
{
    Prefect = 1,
    NotPrefect = 2
};

// Function to read a positive number from the user
int RedPsitiveNumber(string message)
{
    int Number;
    do
    {
        cout << message; // Prompt the user with the given message
        cin >> Number;   // Read the input
    } while (Number < 0); // Ensure the number is positive
    return Number;
}

// Function to check if a given number is a perfect number
enmPrefectNotPerfect CheckPrefectNumber(int Number)
{
    int M = round(Number / 2); // Only check divisors up to half the number
    int SumNubPrime = 0;       // Variable to store the sum of divisors

    // Iterate through potential divisors
    for (int Divisor = 1; Divisor <= M; Divisor++)
    {
        if (Number % Divisor == 0) // Check if Divisor divides Number
            SumNubPrime += Divisor; // Add Divisor to the sum
    }

    // Check if the sum of divisors equals the original number
    if (SumNubPrime == Number)
        return enmPrefectNotPerfect::Prefect; // It is a perfect number

    return enmPrefectNotPerfect::NotPrefect; // It is not a perfect number
}

// Function to print all perfect numbers from 1 to the given number
void PrintPrefectNumber(int Number)
{
    // Iterate through numbers from 1 to the given number
    for (int counter = 1; counter <= Number; counter++)
    {
        // Check if the current number is a perfect number
        if (CheckPrefectNumber(counter) == enmPrefectNotPerfect::Prefect)
            cout << counter << endl; // Print the perfect number
    }
}

int main()
{
    // Prompt the user for a number and print all perfect numbers from 1 to that number
    PrintPrefectNumber(RedPsitiveNumber("Enter a positive number to find all perfect numbers from 1 to N: "));
}

File: Problem solving with Python/Problem04.py
"""
Problem04 |===============================================

Write a program to print all perfect number From 1 to N 
order. 

Input : 
500

Output : 
6
28
496
==========================================================
"""



def read_positive_number(message):
    """
    Function to read a positive number from the user.
    Ensures the input is greater than or equal to zero.
    """
    number = -1  # Initialize with an invalid value
    while number < 0:
        try:
            number = int(input(message))  # Read and convert input to an integer
            if number < 0:
                print("Please enter a positive number.")
        except ValueError:
            print("Invalid input. Please enter an integer.")
    return number

def is_perfect_number(number):
    """
    Function to check if a given number is a perfect number.
    A perfect number is a number whose proper divisors sum to itself.
    """
    m = number // 2  # Only check divisors up to half the number
    sum_divisors = sum(divisor for divisor in range(1, m + 1) if number % divisor == 0)

    return sum_divisors == number  # Returns True if it's a perfect number

def print_perfect_numbers(n):
    """
    Function to print all perfect numbers from 1 to the given number.
    """
    print(f"\nPerfect numbers from 1 to {n}:")
    
    for i in range(1, n + 1):
        if is_perfect_number(i):
            print(i, end=" ")  # Print perfect numbers in a single line

    print()  # Move to a new line after printing

# Main entry point
if __name__ == "__main__":
    num = read_positive_number("Enter a positive number to find all perfect numbers from 1 to N: ")
    print_perfect_numbers(num)

